aten/src/ATen/new_gen.py

```diff
@@ -0,0 +1,871 @@
+        # Currently, we rely directly on the invariant that there are NO
+        # kwarg-only mutating arguments.  If you want to relax this,
+        # we will need a more semantic way of matching that takes
+        # into account return arguments.  In that case, you will have
+        # to manage out_arguments computation a level up, in
+        # FunctionSchema.  See Note [is_out_fn]
+        if parg.annotation is not None and parg.annotation.is_write:
+            if arguments_acc is arguments:
+                pass  # do nothing
+            elif arguments_acc is kwarg_only_arguments:
+                arguments_acc = out_arguments
+        else:
+            assert arguments_acc is not out_arguments
+        arguments_acc.append(parg)
+
+    return arguments, kwarg_only_arguments, out_arguments
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#
+#                           PROCESSING
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+# Conveniently add error context to exceptions raised.  Lets us
+# easily say that an error occurred while processing a specific
+# context.
+@contextlib.contextmanager
+def context(msg: str) -> Iterator[None]:
+    try:
+        yield
+    except Exception as e:
+        # TODO: this does the wrong thing with KeyErorr
+        msg = textwrap.indent(msg, '  ')
+        msg = f'{e.args[0]}\n{msg}' if e.args else msg
+        e.args = (msg,) + e.args[1:]
+        raise
+
+# Represent a source location; used for better error reporting
+@dataclass
+class Location:
+    file: str
+    line: int
+
+    def __str__(self) -> str:
+        return "{}:{}".format(self.file, self.line)
+
+# A custom loader for YAML to let us also keep track of line numbers
+# of each entry in the YAML file
+class LineLoader(Loader):
+    def construct_mapping(self, node, deep=False):  # type: ignore
+        mapping = super().construct_mapping(node, deep=deep)  # type: ignore
+        # Add 1 so line numbering starts at 1
+        mapping['__line__'] = node.start_mark.line + 1
+        return mapping
+
+# Parse native_functions.yaml into a sequence of NativeFunctions
+def parse_native_yaml(path: str) -> List[NativeFunction]:
+    with open(path, 'r') as f:
+        es = yaml.load(f, Loader=LineLoader)
+    assert isinstance(es, list)
+    rs: List[NativeFunction] = []
+    for e in es:
+        assert isinstance(e.get('__line__'), int), e
+        loc = Location(path, e['__line__'])
+        funcs = e.get('func')
+        with context(f'in {loc}:\n  {funcs}'):
+            rs.append(NativeFunction.from_yaml(e, loc))
+    return rs
+
+native_functions = parse_native_yaml('aten/src/ATen/native/native_functions.yaml')
+# pprint.pprint([dataclasses.asdict(f) for f in native_functions])
+
+# TODO: TensorOptions argument detection
+# TODO: Extra enforcement of inplace functions having mutable self
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
```

Comment:
I know this is a prototype, but for the final version, I would propose splitting generation into a separate file from the data structure, probably a separate file for each generated artifact.

Reply:
I agree, having multiple files will be good.